pkgs/development/compilers/llvm/4.0/libc++/default.nix

```diff
@@ -0,0 +1,40 @@
+{ lib, stdenv, fetch, cmake, llvm, libcxxabi, fixDarwinDylibNames, version }:
+
+stdenv.mkDerivation rec {
+  name = "libc++-${version}";
+
+  src = fetch "libcxx" "130clvfffqml8hbnlvr596nfjk18n6ifxab27xl66nxhq99wccfn";
+
+  postUnpack = ''
+    unpackFile ${libcxxabi.src}
+    export LIBCXXABI_INCLUDE_DIR="$PWD/$(ls -d libcxxabi-${version}*)/include"
+  '';
+
+  preConfigure = ''
+    # Get headers from the cxxabi source so we can see private headers not installed by the cxxabi package
+    cmakeFlagsArray=($cmakeFlagsArray -DLIBCXX_CXX_ABI_INCLUDE_PATHS="$LIBCXXABI_INCLUDE_DIR")
+  '';
+
+  patches = lib.optional stdenv.isDarwin ./darwin.patch;
```

Comment:
[Here's the patch in question, as used for 3.9](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/llvm/3.9/libc%2B%2B/darwin.patch)

Hmm.  While looking at the patch to understand it well enough to 'port' it, I noticed a few things:

1) The patch seems to be very close to `substituteInPlace lib/CMakeLists.txt --replace /usr/lib/libc++ '${LIBCXX_LIBCXXABI_LIB_PATH}/libc++'`.  Actually I just checked and on 3.9 at least it appears they're equivalent.

2) One of the paths changed with this patch is originally `/usr/lib/libc++.1.dylib` and afterward `${LIBCXX_LIBCXXABI_LIB_PATH}/libc++.1.dylib`.  This is used for the "install_name".  This seems odd since the currently-building libc++ library will never actually be put in that location (note the libc++abi library path is used).  I'm not sure what "install_name" is used for, but it seems like that this should instead be the path where libc++ will be after installation.  Is this a problem and if so is it worth addressing?

3) The substituteInPlace command above includes the "libc++" suffix duplicated in the patterns in order to avoid changing the path `/usr/lib/libSystem.B.dylib` [which is added to the linker flags on 10.6 apparently](https://github.com/llvm-mirror/libcxx/blob/release_39/lib/CMakeLists.txt#L116).  Is this the intended behavior, and if not should this be changed to a particular libSystem from some other place?

Long story short, I'm thinking drop the patch and just use substituteInPlace and call it good (patch is neater but substituteInPlace is clearer about what is being done)-- the question is whether the paths for "install_name" and libSystem should be changed or not.  I'll start by trying to preserve the 3.9 behavior but would like this addressed/decided before merging this, if possible.